src/python/grpcio/grpc/_cython/_cygrpc/grpc_gevent.pyx.pxi

```diff
@@ -14,418 +14,91 @@
-  rw = ResolveWrapper()
-  rw.c_resolver = r
-  rw.c_host = host
-  rw.c_port = port
-  _spawn_greenlet(socket_resolve_async_python, rw)
-
-cdef grpc_error_handle socket_resolve(const char* host, const char* port,
-                                grpc_resolved_addresses** res) with gil:
-    try:
-      result = gevent_socket.getaddrinfo(host, port)
-      res[0] = tuples_to_resolvaddr(result)
-      return grpc_error_none()
-    except IOError as io_error:
-      return socket_error("getaddrinfo", str(io_error))
-
-###############################
-### timer implementation ######
-###############################
-
-cdef class TimerWrapper:
-  def __cinit__(self, deadline):
-    fork_handlers_and_grpc_init()
-    self.timer = gevent_hub.get_hub().loop.timer(deadline)
-    self.event = None
-
-  def start(self):
-    self.event = gevent_event.Event()
-    self.timer.start(self.on_finish)
-
-  def on_finish(self):
-    grpc_custom_timer_callback(self.c_timer, grpc_error_none())
-    self.timer.stop()
-    g_event.set()
-
-  def stop(self):
-    self.event.set()
-    self.timer.stop()
-
-  def __dealloc__(self):
-    grpc_shutdown()
-
-cdef void timer_start(grpc_custom_timer* t) with gil:
-  timer = TimerWrapper(t.timeout_ms / 1000.0)
-  timer.c_timer = t
-  t.timer = <void*>timer
-  timer.start()
-
-cdef void timer_stop(grpc_custom_timer* t) with gil:
-  time_wrapper = <object>t.timer
-  time_wrapper.stop()
-
-###############################
-### pollset implementation ###
-###############################
-
-cdef void init_loop() with gil:
-  pass
-
-cdef void destroy_loop() with gil:
-  g_pool.join()
-
-cdef void kick_loop() with gil:
-  g_event.set()
-
-def _run_loop(timeout_ms):
-  timeout = timeout_ms / 1000.0
-  if timeout_ms > 0:
-    try:
-      g_event.wait(timeout)
-    finally:
-      g_event.clear()
-
-cdef grpc_error_handle run_loop(size_t timeout_ms) with gil:
-  try:
-    _run_loop(timeout_ms)
-    return grpc_error_none()
-  except BaseException:
-    exc_info = sys.exc_info()
-    # Avoid running any Python code after setting the exception
-    cpython.PyErr_SetObject(exc_info[0], exc_info[1])
-    return GRPC_ERROR_CANCELLED
-
-###############################
-### Initializer ###############
-###############################
+import atexit
+import errno
+import sys
 
-cdef grpc_socket_vtable gevent_socket_vtable
-cdef grpc_custom_resolver_vtable gevent_resolver_vtable
-cdef grpc_custom_timer_vtable gevent_timer_vtable
-cdef grpc_custom_poller_vtable gevent_pollset_vtable
+gevent_hub = None
+g_gevent_pool = None
+g_gevent_threadpool = None
+g_gevent_activated = False
+
+
+cdef queue[void*] g_greenlets_to_run
+cdef condition_variable g_greenlets_cv
+cdef mutex g_greenlets_mu
+cdef bint g_shutdown_greenlets_to_run_queue = False
+
+
+cdef _submit_to_greenlet_queue(object cb, tuple args):
+  cdef tuple to_call = (cb,) + args
+  cdef unique_lock[mutex]* lk
```

Comment:
I wonder if we don't support true multi-threading, maybe we don't need locks in this file?

Reply:
I see, it's for metadata plugin.

Reply:
Right. This does involve actual threads and therefore does require mutexes.